app/services/report_composer.rb

```diff
@@ -0,0 +1,105 @@
+class ReportComposer
+  include ActiveModel::Model
+
+  class InputValue
+    include ActiveModel::Model
+
+    attr_accessor :value, :template_input
+
+    validates :value, :presence => true, :if => proc { |v| v.template_input.required? || v.value.present? }
+
+    validates :value, :inclusion => { :in => proc { |v| v.template_input.options_array } },
+              :if => proc { |v| v.template_input.input_type == 'user' && v.template_input.options_array.present? }
+  end
+
+  class UiParams
+    attr_reader :ui_params
```

Comment:
If you name this `params` in both `UiParams`  and `ApiParams` they could share a common base class with a common `attr_reader` and `initialize`. Not sure if that's worth it, though.

Reply:
These are internal helper objects that most likely won't be used outside, while now it looks they are very similar, one day there's might be another that takes existing persisted composer and build a new composer instance out of it (like in REX) and then params parameter would not make sense as the only input will be the existing composer. Also it helps me to understand that these are ui/api params, which differs in structure. If you don't mind, I'd keep it as is.